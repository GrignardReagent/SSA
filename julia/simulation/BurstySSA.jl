module BurstySSA

using Distributions: Geometric
using Catalyst
using JumpProcesses
using DifferentialEquations
using DataFrames
using Random
using Base.Threads

# symbolic registration must be evaluated at the top level 
@register_symbolic Geometric(b)

export simulate_bursty_model # this can be used directly when the module is imported

"""
    simulate_bursty_model(parameter_sets, time_points, ntrajectories; num_cores=nthreads())

Simulate mRNA counts from the 2-state bursty model via Gillespie SSA.

- `parameter_sets`: a Vector of Dicts with keys:
    "k", "b", "d", and optional "label".
- `time_points`: Vector of times at which to save M counts (e.g. `collect(0.0:1.0:100.0)`).
- `ntrajectories`: number of trajectories per parameter set.
- `num_cores`: number of threads to use (defaults to `Threads.nthreads()`).

Returns a `DataFrame` with columns: `:label, :time_t1, :time_t2, ...`.
"""
function simulate_bursty_model(parameter_sets, time_points, traj::Int; num_cores::Int=nthreads())
    
    @parameters b
    r = rand(Geometric(1/b)) + 1
    # the network
    bursty_model = @reaction_network bursty_model begin
        k, G --> G + $r*P
        d, P --> 0 
        end

    # initial molecule counts
    u0 = [:G => 1, :P => 0]
    tspan = (time_points[1], time_points[end])

    rows = Vector{Vector{Any}}()  # Any to allow String or Int labels

    @threads for ps in parameter_sets
        # parameters in the order rn.ps (Catalyst guarantees param order in rn.ps)
        params = Dict(
            :k => ps["k"],
            :b => ps["b"],
            :d => ps["d"],
        )

        label = Int32(get(ps, "label", 0))

        if traj == 1
            # single traj
            row = run_single(bursty_model, u0, tspan, params, time_points, label)
            push!(rows, row)
        else
            # multiple trajs, use ensemble
            ensemble_results = run_ensemble(bursty_model, u0, tspan, params, time_points, label, traj)
            append!(rows, ensemble_results)
        end
    end

    colsyms = [:label; Symbol.("time_" .* string.(time_points))]

    # Convert rows to a proper matrix format
    data_matrix = Matrix{Any}(undef, length(rows), length(colsyms))
    for (i, row) in enumerate(rows)
        data_matrix[i, :] = row
    end

    df = DataFrame(data_matrix, colsyms)
    return df
end

# One SSA trajectory, returns a row: [label, P(t1), P(t2), ...]
function run_single(bursty_model, u0, tspan, params, time_points, label)
    jinput = JumpInputs(bursty_model, u0, tspan, params)
    jprob  = JumpProblem(jinput; save_positions=(false, false)) # Disable saving at jump positions
    sol       = solve(jprob, SSAStepper(), saveat = time_points)
    p_counts  = Array(sol)[2, :]  # species order is [G, P], so we choose the 2nd row
    return [label; p_counts...]  # label + P counts
end

# run an ensemble of trajectories
function run_ensemble(bursty_model, u0, tspan, params, time_points, label, traj)
    jinput = JumpInputs(bursty_model, u0, tspan, params)
    jprob  = JumpProblem(jinput; save_positions=(false, false))
    eprob = EnsembleProblem(jprob)
    esol = solve(eprob, SSAStepper(); trajectories=traj, saveat=time_points)
    
    # Extract P counts for each trajectory
    ensemble_array = Array(esol)  # shape: (species, time_points, trajectories)
    p_counts_all = ensemble_array[2, :, :]  # P is the 2nd species
    
    # Create one row per trajectory
    rows = Vector{Vector{Any}}(undef, traj)
    for i in 1:traj
        p_counts = p_counts_all[:, i]  # P counts for trajectory i
        rows[i] = [label; p_counts...]
    end
    
    return rows
end

end # module