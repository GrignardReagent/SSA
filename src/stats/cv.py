#!/usr/bin/python
from stats.mean import calculate_mean_from_params
from stats.variance import calculate_variance_from_params
import numpy as np

def calculate_cv(variance, mean):
    """
    Calculate the coefficient of variation (CV).
    
    Args:
        variance (float): The variance.
        mean (float): The mean.
        
    Returns:
        float: The calculated coefficient of variation.
    """
    import numpy as np
    cv = np.sqrt(variance) / mean
    return cv

def calculate_cv_from_params(rho, d, sigma_b, sigma_u):
    """
    Calculate the coefficient of variation (CV) from parameters.
    Equation: CV = sqrt(Var)/Mean
    
    Parameters:
        rho (float): The value of rho.
        d (float): The value of d.
        sigma_b (float): The value of sigma_b.
        sigma_u (float): The value of sigma_u.
        
    Returns:
        float: The calculated coefficient of variation.
    """
    mu = calculate_mean_from_params(rho, d, sigma_b, sigma_u)
    var = calculate_variance_from_params(rho, d, sigma_b, sigma_u)
    return np.sqrt(var) / mu