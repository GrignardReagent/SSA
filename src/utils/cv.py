#!/usr/bin/python

def calculate_cv(variance, mean):
    """
    Calculate the coefficient of variation (CV).
    
    Args:
        variance (float): The variance.
        mean (float): The mean.
        
    Returns:
        float: The calculated coefficient of variation.
    """
    import numpy as np
    cv = np.sqrt(variance) / mean
    return cv
