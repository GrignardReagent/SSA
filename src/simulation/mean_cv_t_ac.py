#!/usr/bin/python

import numpy as np
from scipy.optimize import root_scalar
from utils.biological import check_biological_appropriateness
from stats.mean import calculate_mean_from_params
from stats.variance import calculate_variance_from_params
from stats.autocorrelation import calculate_ac_from_params
import warnings

# # Rescale parameters
# rho_tilda = rho / (sigma_b + sigma_u)
# d_tilda = d / (sigma_b + sigma_u)
# sigma_b_tilda = sigma_b / (sigma_b + sigma_u)
# sigma_u_tilda = sigma_u / (sigma_b + sigma_u)
# # t_ac_tilda is the only one that's multiplied by the scaling factor 
# t_ac_tilda = t_ac * (sigma_b + sigma_u)

# Compute rescaled parameters for a fixed ``sigma_sum``.
#
# This helper solves the mean, CV and autocorrelation time equations in the
# rescaled space assuming a known sum of switching rates ``sigma_sum`` (default = 1).
# It is intentionally kept private; the public :func:`find_tilda_parameters`
# routine wraps this function and automatically searches for an appropriate
# ``sigma_sum``.
def find_tilda_parameters(
    mu_target: float,
    t_ac_target: float,
    cv_target: float,
    sigma_sum: float = 1.0,
    ac_target: float = np.exp(-1),
    res_limit: float = 1e-3,
    lower: float = 1e-3,
    upper: float = 1e3,
    check_biological: bool = False,
    max_fano_factor: float = 20,
    min_fano_factor: float = 1.0,
    max_cv: float = 5,
):
    """Solve for ``rho``, ``d``, ``sigma_b`` and ``sigma_u`` given ``sigma_sum``.

    The equations are expressed in a rescaled space where all parameters are divided by ``sigma_sum`` to simplify the algebra.  

    Parameters
    ----------
    mu_target : float
        Target mean of the system.
    t_ac_target : float
        Target autocorrelation time.
    cv_target : float
        Desired coefficient of variation.
    sigma_sum : float, optional
        The sum ``sigma_b + sigma_u`` used to scale the system. Default to be 1.0. Changing this might lead to unexpected bahaviours!
    ac_target : float, optional
        Target autocorrelation value at ``t_ac_target``. Defaults to ``exp(-1)``.
    res_limit : float, optional
        Residual limit for root-finding. Defaults to ``1e-3``. This is used to check if the solution found is valid or not.
    lower : float, optional
        Lower bound for the root-finding search space for d_tilda. Defaults to ``1e-3``
    upper : float, optional
        Upper bound for the root-finding search space for d_tilda. Defaults to ``1e3``.
    check_biological : bool, optional
        If True, checks if the solution is biologically appropriate. Defaults to False.
    max_fano_factor : float, optional
        Maximum allowed Fano factor for the solution. Defaults to 20.
    min_fano_factor : float, optional
        Minimum allowed Fano factor for the solution. Defaults to 1.0.
    max_cv : float, optional
        Maximum allowed coefficient of variation for the solution. Defaults to 5.

    Returns
    -------
    tuple
        ``(rho, d, sigma_b, sigma_u)`` that reproduce the requested targets.

    Raises
    ------
    ValueError
        If no valid solution for ``d`` can be found.
    """
    
    # Basic feasibility checks
    if mu_target <= 0:
        raise ValueError("mu_target must be > 0.")
    if cv_target <= 0:
        raise ValueError("cv_target must be > 0.")
    if t_ac_target <= 0:
        raise ValueError("t_ac_target must be > 0 (AC(0)=1 carries no info).")
    if sigma_sum <= 0:
        raise ValueError("sigma_sum (Σ) must be > 0.")

    
    # using equation B, we find v via mu_target & cv_target^2
    v = (mu_target * (cv_target ** 2)) - 1
    
    # check if v > 0, if not, there's no solution
    if v <= 0:
        raise ValueError(f"Invalid parameters: v = (mu_target * cv_target ** 2) - 1 must be positive, got v = {v}. Reconsider mu_target and cv_target choices.")
    
    # t_ac_tilda = (sigma_b + sigma_u) * t; 
    t_ac_tilda = t_ac_target * sigma_sum
    
    # scaled lag: warn if far from 1 (conditioning)
    if t_ac_tilda < 1e-3 or t_ac_tilda > 1e3:
        warnings.warn(
            f"Scaled lag t̃ = Σ·t_ac = {t_ac_tilda:.3g} is ill-conditioned; "
            f"consider Σ≈1/t_ac_target so t̃≈1."
        )
    
    # find d_tilda from equation C, via t_ac_tilda, v
    def scaled_ac_equation(d_tilda):
        '''
        AC(t_ac_tilda), rescaled equation of AC(t) to find d_tilda:
        '''
        
        scaled_ACmRNA_eq = ((d_tilda * v * np.exp(- t_ac_tilda) + (d_tilda - v - 1) * np.exp(- d_tilda * t_ac_tilda)) / ((d_tilda -1) * (v + 1)))

        
        return float(scaled_ACmRNA_eq - ac_target)
    
    ################ Use root scalar to find d_tilda, AC(t_ac_tilda) - AC_target = 0  is a root-finding problem, so root_scalar is more appropriate ##################
    
    # the AC(t_ac_tilda) formula has the denominator (d_tilda -1)(v + 1), and d_tilda = 1 is undefined, so the search space needs to exclude 1.0; multiple decimal places to bracket away from 1.0
    candidates = [(lower, 0.999), (1.001, upper)] if (lower < 1.0 < upper) else [(lower, upper)]  # several decimal places are intentionally used here to avoid values close to 1.0
    
    d_tilda = None
    for a, b in candidates:
        try:
            result = root_scalar(scaled_ac_equation, bracket=[a, b],method='brentq', # Brent's method for root finding, generally considered the best for this. 
                                 xtol=1e-10, rtol=1e-10, maxiter=200)
            
            if result.converged:
                d_tilda_solution = result.root
                
                # error trap for no solution found for d_tilda_solution
                # 1. checks that absolute value of residue is less than res_limit
                # 2. checks that d_tilda_solution is not None
                # 3. checks that the value of d_tilda_solution is not close to the upper bounds plus/minus the absolute tolerance (atol)
                residue = scaled_ac_equation(d_tilda_solution)
                if abs(residue) > res_limit or d_tilda_solution is None or np.isclose(d_tilda_solution, b, atol=1e-4):
                    raise  ValueError(f"⚠️No valid solution found for parameter d_tilda: residual={residue:.2e}, d_tilda={d_tilda_solution:.4f}")
                
                d_tilda = d_tilda_solution
                
        except ValueError:
            pass
    if d_tilda is None: 
        raise ValueError(f'Could not find a valid solution for d_tilda using root_scalar method.')
    
    ################ Use root scalar to find d_tilda ##################
            
    # find rho_tilda from mu_target, d_tilda and v; by rearranging Equation A, definition of v in terms of tilda_params.
    #TODO: explanation of maths...
    rho_tilda = mu_target * d_tilda + (d_tilda + 1) * v
    
    # find sigma_b_tilda using solved parameters
    sigma_b_tilda = mu_target * d_tilda / (mu_target * d_tilda + (d_tilda + 1) * v)

    # find sigma_u_tilda using solved parameters
    sigma_u_tilda = (d_tilda + 1) * v / (mu_target * d_tilda + (d_tilda + 1) * v)
    
    # scaling back to original param space
    sigma_b = sigma_b_tilda * sigma_sum
    sigma_u = sigma_u_tilda * sigma_sum
    rho = rho_tilda * sigma_sum
    d = d_tilda * sigma_sum
    
    # optional biological check
    if check_biological:
        variance_target = (cv_target * mu_target) ** 2
        is_appropriate = check_biological_appropriateness(
            variance_target, mu_target, max_fano_factor, min_fano_factor, max_cv
        )
        if not is_appropriate:
            raise ValueError(
                "Solution is not biologically appropriate. "
                f"Consider adjusting target parameters: mu={mu_target}, cv={cv_target}, "
                f"which gives variance={variance_target:.2f}"
            )
        
    # back-substitution verification against original formulas
    mu_analytical = calculate_mean_from_params(rho, d, sigma_b, sigma_u)
    var_analytical = calculate_variance_from_params(rho, d, sigma_b, sigma_u)
    cv_analytical = np.sqrt(var_analytical) / mu_analytical
    ac_analytical = calculate_ac_from_params(rho, d, sigma_b, sigma_u, t_ac_target)
    
    # relative/absolute residuals
    rel = lambda x, y: abs(x - y) / max(1.0, abs(y)) # behaves like a relative error when abs(y) >= 1 and like an absolute error when abs(y) < 1. In contrastm, using just abs(y) in the denominator blows up (or becomes overly strict) when y is tiny or zero.
    mu_err = rel(mu_analytical, mu_target)
    cv_err = rel(cv_analytical, cv_target)
    ac_err = abs(ac_analytical - ac_target)
    if mu_err > res_limit or cv_err > res_limit or ac_err > res_limit:
        raise ValueError(
            "Back-substitution check failed: "
            f"μ_err={mu_err:.2e}, CV_err={cv_err:.2e}, AC_err={ac_err:.2e}. "
            "Try adjusting Σ so that t̃≈1, or relax targets."
        )
    return rho, d, sigma_b, sigma_u
