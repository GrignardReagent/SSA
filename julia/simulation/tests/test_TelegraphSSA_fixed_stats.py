#!/usr/bin/env python3

# import the simulation functions
import os
os.environ.setdefault("PYTHON_JULIACALL_HANDLE_SIGNALS", "yes")
os.environ["JULIA_NUM_THREADS"] = "8"  # or set to desired number of threads
from juliacall import Main as jl
import numpy as np
import pandas as pd 
from simulation.mean_cv_t_ac import find_tilda_parameters
# import the stats functions
from stats.mean import calculate_mean
from stats.variance import calculate_variance
from stats.cv import calculate_cv
from stats.autocorrelation import calculate_autocorrelation, calculate_ac_time_interp1d
# plotting modules
import matplotlib.pyplot as plt
from visualisation.plots import plot_mRNA_trajectory, plot_mRNA_dist

'''
Test script: testing whether the julia implementation of the telegraph SSA can produce trajectorie s matching a set of prescribed stats.
'''

def run_simulation_test(mu_target, t_ac_target, cv_target, size=1000):
    """Run a single simulation test for given target statistics."""
    ###############################################################
    # Simulation
    ###############################################################
    print(f"\n{'='*60}")
    print(f"Testing: μ={mu_target}, t_ac={t_ac_target}, CV={cv_target}")
    print(f"{'='*60}")
    
    # Get the parameters using the scaled equations
    rho, d, sigma_b, sigma_u = find_tilda_parameters(mu_target, t_ac_target, cv_target)
    print(f"Parameters: ρ={rho:.6f}, d={d:.6f}, σ_b={sigma_b:.6f}, σ_u={sigma_u:.6f}")
    
    # Prepare parameter sets
    parameter_sets = [{
        'sigma_b': sigma_b,
        'sigma_u': sigma_u,
        'rho': rho,
        'd': d,
        'label': 0
    }]
    
    # Set time points - longer simulation for better autocorrelation estimation
    time_points = np.arange(0, 10_000, 1.0)
    
    # Python → Julia conversion handled automatically
    jl.parameter_sets = parameter_sets
    jl.time_points = time_points
    
    # Run the simulation in Julia
    jl.seval(f'df = simulate_telegraph_model(parameter_sets, time_points, {size})')
    
    # Convert Julia DataFrame directly to Python
    labels = np.array(jl.seval('Int64.(df.label)'))
    counts_matrix = np.array(jl.seval('Int64.(Matrix(df[:, Not(:label)]))'))
    
    # Create pandas DataFrame
    df_labels = pd.DataFrame(labels, columns=['label'])
    df_counts = pd.DataFrame(counts_matrix, columns=[f"time_{ti}" for ti in time_points])
    df = pd.concat([df_labels, df_counts], axis=1)
    
    ###############################################################
    # Now calculate the stats from the simulated data
    ###############################################################
    mean_observed = calculate_mean(df_counts.T, parameter_sets, use_steady_state=False) # note the df is transposed
    variance_observed = calculate_variance(df_counts.T, parameter_sets, use_steady_state=False)
    cv_observed = calculate_cv(variance_observed, mean_observed)
    
    # Calculate autocorrelation
    autocorr_results = calculate_autocorrelation(df)
    ac_mean = autocorr_results['stress_ac'].mean(axis=0)
    lags = autocorr_results['stress_lags']
    t_ac_observed = calculate_ac_time_interp1d(ac_mean, lags)

    # Calculate errors
    mean_error_pct = 100 * abs(np.mean(mean_observed) - mu_target) / mu_target
    cv_error_pct = 100 * abs(np.mean(cv_observed) - cv_target) / cv_target
    t_ac_error_pct = 100 * abs(t_ac_observed - t_ac_target) / t_ac_target

    ###############################################################
    # Reporting 
    ###############################################################
    # Print results
    print(f"Mean: Target = {mu_target}, Observed = {np.mean(mean_observed):.3f} (Error: {mean_error_pct:.1f}%)")
    print(f"CV: Target = {cv_target}, Observed = {np.mean(cv_observed):.3f} (Error: {cv_error_pct:.1f}%)")
    print(f"Variance: Observed = {np.mean(variance_observed):.3f}")
    print(f"AC Time: Target = {t_ac_target}, Observed = {t_ac_observed:.3f} (Error: {t_ac_error_pct:.1f}%)")
    
    ###############################################################
    # Plotting 
    ###############################################################
    # Create output directory if it doesn't exist
    output_dir = "test_telegraph_fixed_stats_results"
    os.makedirs(output_dir, exist_ok=True)
    
    # Create and save traj plot
    fig, ax = plot_mRNA_trajectory(parameter_sets, time_points, df_counts)  
    # Save figure with descriptive filename
    traj_filename = f"{output_dir}/telegraph_test_mu{mu_target}_tac{t_ac_target}_cv{cv_target}_traj.png"
    fig.savefig(traj_filename, dpi=300, bbox_inches='tight')
    plt.close(fig)  # Close to free memory
    print(f"Trajectory plot saved to: {traj_filename}")

    # Plot data distribution
    fig, ax = plot_mRNA_dist(parameter_sets, df_counts)
    # Save figure with descriptive filename
    dist_filename = f"{output_dir}/telegraph_test_mu{mu_target}_tac{t_ac_target}_cv{cv_target}_dist.png"
    fig.savefig(dist_filename, dpi=300, bbox_inches='tight')
    plt.close(fig)  # Close to free memory
    print(f"Data distribution plot saved to: {dist_filename}")

    # Plot autocorrelation
    plt.figure(figsize=(10, 6))
    # Find indices for positive lags only (autocorrelation is symmetric)
    positive_mask = lags >= 0
    positive_lags = lags[positive_mask]
    positive_mean_ac = ac_mean[positive_mask]
    # Limit to first 200 points
    max_lag_points = min(200, len(positive_lags))
    plot_lags = positive_lags[:max_lag_points]
    plot_ac = positive_mean_ac[:max_lag_points]
    # Plot stress condition (zoomed in)
    plt.plot(plot_lags, plot_ac, color='blue', label=f'AC Time: {t_ac_observed:.2f}')
    # Plot the AC time lines, show the values (only if within the plot range)
    plt.axhline(y=1/np.e, color='gray', linestyle='--', label='AC = 1/e')
    if t_ac_observed <= plot_lags[-1]:  # Only show vertical line if it's within the plot range
        plt.axvline(x=t_ac_observed, color='blue', linestyle='--', alpha=0.7)
    plt.title('Autocorrelation (First 200 Lag Points)')
    plt.xlabel('Lag')
    plt.ylabel('Autocorrelation')
    plt.xlim(0, plot_lags[-1])  # Set x-axis limits explicitly
    plt.grid(True, alpha=0.3)
    plt.legend()
    plt.tight_layout()
    # saving the figure
    tac_filename = f"{output_dir}/telegraph_test_mu{mu_target}_tac{t_ac_target}_cv{cv_target}_tac.png"
    plt.savefig(tac_filename, dpi=300, bbox_inches='tight')
    plt.close()  # Close to free memory
    print(f"Autocorrelation plot saved to: {tac_filename}")
    
    
    return {
        'mu_target': mu_target,
        't_ac_target': t_ac_target,
        'cv_target': cv_target,
        'mu_observed': np.mean(mean_observed),
        'cv_observed': np.mean(cv_observed),
        't_ac_observed': t_ac_observed,
        'variance_observed': np.mean(variance_observed),
        'mean_error_pct': mean_error_pct,
        'cv_error_pct': cv_error_pct,
        't_ac_error_pct': t_ac_error_pct,
        'parameters': {'rho': rho, 'd': d, 'sigma_b': sigma_b, 'sigma_u': sigma_u}
    }

###############################################################
# Main execution
###############################################################

# Initialize Julia environment once
print("Initializing Julia environment...")
jl.seval('using Pkg; Pkg.activate("/home/ianyang/stochastic_simulations/julia"); Pkg.instantiate()')
jl.seval('using DataFrames, NPZ, Base.Threads')
jl.include("/home/ianyang/stochastic_simulations/julia/simulation/TelegraphSSA.jl")
jl.seval('using .TelegraphSSA')

# sanity check: how many threads did we get?
nthreads = int(jl.seval('nthreads()'))
print(f"Julia nthreads = {nthreads}")

# Define test targets
mu_targets = [10, 100, 1000]
t_ac_targets = [1, 5, 10]
cv_targets = [0.5, 1.0, 2.0]

# Run all combinations
results = []
total_tests = len(mu_targets) * len(t_ac_targets) * len(cv_targets)
current_test = 0

for mu_target in mu_targets:
    for t_ac_target in t_ac_targets:
        for cv_target in cv_targets:
            current_test += 1
            print(f"\nRunning test {current_test}/{total_tests}")
            
            try:
                result = run_simulation_test(mu_target, t_ac_target, cv_target)
                results.append(result)
            except Exception as e:
                print(f"Error in test μ={mu_target}, t_ac={t_ac_target}, CV={cv_target}: {e}")
                continue

# Create summary table
print(f"\n{'='*80}")
print("SUMMARY OF ALL TESTS")
print(f"{'='*80}")

summary_df = pd.DataFrame(results)
if not summary_df.empty:
    print(summary_df[['mu_target', 't_ac_target', 'cv_target', 
                     'mu_observed', 'cv_observed', 't_ac_observed',
                     'mean_error_pct', 'cv_error_pct', 't_ac_error_pct']].round(3))

    # Save summary to CSV
    summary_df.to_csv("test_telegraph_fixed_stats_results/summary_results.csv", index=False)
    print(f"\nSummary saved to: test_telegraph_fixed_stats_results/summary_results.csv")

    # Print overall statistics
    print(f"\nOverall Error Statistics:")
    print(f"Mean Error: {summary_df['mean_error_pct'].mean():.2f}% ± {summary_df['mean_error_pct'].std():.2f}%")
    print(f"CV Error: {summary_df['cv_error_pct'].mean():.2f}% ± {summary_df['cv_error_pct'].std():.2f}%")
    print(f"AC Time Error: {summary_df['t_ac_error_pct'].mean():.2f}% ± {summary_df['t_ac_error_pct'].std():.2f}%")

    # Check for any large errors
    large_errors = summary_df[(summary_df['mean_error_pct'] > 10) | 
                             (summary_df['cv_error_pct'] > 15) | 
                             (summary_df['t_ac_error_pct'] > 20)]
    if not large_errors.empty:
        print(f"\n⚠️ Tests with large errors (>10% mean, >15% CV, >20% AC time):")
        print(large_errors[['mu_target', 't_ac_target', 'cv_target', 
                           'mean_error_pct', 'cv_error_pct', 't_ac_error_pct']].round(2))
else:
    print("No successful tests completed.")

print(f"\nAll tests completed. Results saved in 'telegraph_test_results' directory.")
