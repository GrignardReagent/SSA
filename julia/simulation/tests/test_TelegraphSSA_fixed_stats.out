Initializing Julia environment...
Julia nthreads = 12

Running test 1/27

============================================================
Testing: μ=10, t_ac=1, CV=0.5
============================================================
Parameters: ρ=30.250000, d=2.500000, σ_b=0.826446, σ_u=0.173554
Mean: Target = 10, Observed = 10.000 (Error: 0.0%)
CV: Target = 0.5, Observed = 0.500 (Error: 0.0%)
Variance: Observed = 25.000
AC Time: Target = 1, Observed = 1.000 (Error: 0.0%)
Trajectory plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu10_tac1_cv0.5_traj.png
Data distribution plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu10_tac1_cv0.5_dist.png
Autocorrelation plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu10_tac1_cv0.5_tac.png

Running test 2/27

============================================================
Testing: μ=10, t_ac=1, CV=1.0
============================================================
Parameters: ρ=199.000000, d=10.000000, σ_b=0.502513, σ_u=0.497487
Mean: Target = 10, Observed = 10.002 (Error: 0.0%)
CV: Target = 1.0, Observed = 1.000 (Error: 0.0%)
Variance: Observed = 99.996
AC Time: Target = 1, Observed = 1.000 (Error: 0.0%)
Trajectory plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu10_tac1_cv1.0_traj.png
Data distribution plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu10_tac1_cv1.0_dist.png
Autocorrelation plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu10_tac1_cv1.0_tac.png

Running test 3/27

============================================================
Testing: μ=10, t_ac=1, CV=2.0
============================================================
Parameters: ρ=1999.000000, d=40.000000, σ_b=0.200100, σ_u=0.799900
Mean: Target = 10, Observed = 10.005 (Error: 0.0%)
CV: Target = 2.0, Observed = 2.000 (Error: 0.0%)
Variance: Observed = 400.197
AC Time: Target = 1, Observed = 1.000 (Error: 0.0%)
Trajectory plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu10_tac1_cv2.0_traj.png
Data distribution plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu10_tac1_cv2.0_dist.png
Autocorrelation plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu10_tac1_cv2.0_tac.png

Running test 4/27

============================================================
Testing: μ=10, t_ac=5, CV=0.5
============================================================
Parameters: ρ=4.177016, d=0.232784, σ_b=0.557297, σ_u=0.442703
Mean: Target = 10, Observed = 9.988 (Error: 0.1%)
CV: Target = 0.5, Observed = 0.501 (Error: 0.1%)
Variance: Observed = 24.990
AC Time: Target = 5, Observed = 4.996 (Error: 0.1%)
Trajectory plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu10_tac5_cv0.5_traj.png
Data distribution plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu10_tac5_cv0.5_dist.png
Autocorrelation plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu10_tac5_cv0.5_tac.png

Running test 5/27

============================================================
Testing: μ=10, t_ac=5, CV=1.0
============================================================
Parameters: ρ=13.784452, d=0.251813, σ_b=0.182679, σ_u=0.817321
Mean: Target = 10, Observed = 9.999 (Error: 0.0%)
CV: Target = 1.0, Observed = 1.000 (Error: 0.0%)
Variance: Observed = 100.004
AC Time: Target = 5, Observed = 5.002 (Error: 0.0%)
Trajectory plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu10_tac5_cv1.0_traj.png
Data distribution plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu10_tac5_cv1.0_dist.png
Autocorrelation plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu10_tac5_cv1.0_tac.png

Running test 6/27

============================================================
Testing: μ=10, t_ac=5, CV=2.0
============================================================
Parameters: ρ=51.585813, d=0.256853, σ_b=0.049791, σ_u=0.950209
Mean: Target = 10, Observed = 10.026 (Error: 0.3%)
CV: Target = 2.0, Observed = 1.998 (Error: 0.1%)
Variance: Observed = 401.246
AC Time: Target = 5, Observed = 4.996 (Error: 0.1%)
Trajectory plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu10_tac5_cv2.0_traj.png
Data distribution plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu10_tac5_cv2.0_dist.png
Autocorrelation plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu10_tac5_cv2.0_tac.png

Running test 7/27

============================================================
Testing: μ=10, t_ac=10, CV=0.5
============================================================
Parameters: ρ=2.729779, d=0.106937, σ_b=0.391743, σ_u=0.608257
Mean: Target = 10, Observed = 9.997 (Error: 0.0%)
CV: Target = 0.5, Observed = 0.500 (Error: 0.0%)
Variance: Observed = 24.981
AC Time: Target = 10, Observed = 9.973 (Error: 0.3%)
Trajectory plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu10_tac10_cv0.5_traj.png
Data distribution plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu10_tac10_cv0.5_dist.png
Autocorrelation plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu10_tac10_cv0.5_tac.png

Running test 8/27

============================================================
Testing: μ=10, t_ac=10, CV=1.0
============================================================
Parameters: ρ=11.101554, d=0.110608, σ_b=0.099633, σ_u=0.900367
Mean: Target = 10, Observed = 9.983 (Error: 0.2%)
CV: Target = 1.0, Observed = 0.999 (Error: 0.1%)
Variance: Observed = 99.395
AC Time: Target = 10, Observed = 9.973 (Error: 0.3%)
Trajectory plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu10_tac10_cv1.0_traj.png
Data distribution plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu10_tac10_cv1.0_dist.png
Autocorrelation plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu10_tac10_cv1.0_tac.png

Running test 9/27

============================================================
Testing: μ=10, t_ac=10, CV=2.0
============================================================
Parameters: ρ=44.465780, d=0.111547, σ_b=0.025086, σ_u=0.974914
Mean: Target = 10, Observed = 10.055 (Error: 0.5%)
CV: Target = 2.0, Observed = 1.989 (Error: 0.5%)
Variance: Observed = 400.049
AC Time: Target = 10, Observed = 9.961 (Error: 0.4%)
Trajectory plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu10_tac10_cv2.0_traj.png
Data distribution plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu10_tac10_cv2.0_dist.png
Autocorrelation plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu10_tac10_cv2.0_tac.png

Running test 10/27

============================================================
Testing: μ=100, t_ac=1, CV=0.5
============================================================
Parameters: ρ=3124.000000, d=25.000000, σ_b=0.800256, σ_u=0.199744
Mean: Target = 100, Observed = 99.996 (Error: 0.0%)
CV: Target = 0.5, Observed = 0.500 (Error: 0.0%)
Variance: Observed = 2499.914
AC Time: Target = 1, Observed = 1.000 (Error: 0.0%)
Trajectory plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu100_tac1_cv0.5_traj.png
Data distribution plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu100_tac1_cv0.5_dist.png
Autocorrelation plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu100_tac1_cv0.5_tac.png

Running test 11/27

============================================================
Testing: μ=100, t_ac=1, CV=1.0
============================================================
Parameters: ρ=19999.000000, d=100.000000, σ_b=0.500025, σ_u=0.499975
Mean: Target = 100, Observed = 99.980 (Error: 0.0%)
CV: Target = 1.0, Observed = 1.000 (Error: 0.0%)
Variance: Observed = 9998.418
AC Time: Target = 1, Observed = 1.000 (Error: 0.0%)
Trajectory plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu100_tac1_cv1.0_traj.png
Data distribution plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu100_tac1_cv1.0_dist.png
Autocorrelation plot saved to: test_telegraph_fixed_stats_results/telegraph_test_mu100_tac1_cv1.0_tac.png

Running test 12/27

============================================================